File: example/Lab05_2.py
"""
Lab05_2.py - Mã hóa/Giải mã tập tin với RSA
"""

from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_OAEP
import time
import os
import sys

class RSAFileEncryption:
    def __init__(self):
        self.key_pair = None
        self.public_key = None
        self.private_key = None
    
    def generate_keys(self, key_size=2048):
        print(f"Đang tạo cặp khóa RSA {key_size} bits...")
        start = time.time()
        self.key_pair = RSA.generate(key_size)
        print(f"Tạo khóa hoàn tất trong {time.time() - start:.2f} giây")
        
        self.private_key = self.key_pair
        self.public_key = self.key_pair.publickey()
    
    def save_keys(self, base_filename):
        with open(f"{base_filename}.key", 'wb') as f:
            f.write(self.private_key.export_key('DER'))
        

        with open(f"{base_filename}.pub", 'wb') as f:
            f.write(self.public_key.export_key('DER'))
        
        print(f"Đã lưu khóa:")
        print(f"  Private key: {base_filename}.key")
        print(f"  Public key: {base_filename}.pub")
    
    def load_public_key(self, filename):
        """
        Tải public key từ file
        """
        with open(filename, 'rb') as f:
            key_data = f.read()
        self.public_key = RSA.import_key(key_data)
        print(f"Đã tải public key từ {filename}")
    
    def load_private_key(self, filename):
        """
        Tải private key từ file
        """
        with open(filename, 'rb') as f:
            key_data = f.read()
        self.private_key = RSA.import_key(key_data)
        print(f"Đã tải private key từ {filename}")
    
    def encrypt_file(self, input_file, output_file):
        """
        Mã hóa tập tin với RSA
        RSA chỉ có thể mã hóa dữ liệu nhỏ, nên sẽ chia file thành các block
        
        Args:
            input_file: File đầu vào
            output_file: File đầu ra (đã mã hóa)
        """
        if self.public_key is None:
            raise ValueError("Chưa có public key. Hãy tạo hoặc tải khóa trước.")
        
        start_time = time.time()
        cipher = PKCS1_OAEP.new(self.public_key)
        key_size_bytes = self.public_key.size_in_bytes()
        max_block_size = key_size_bytes - 2*20 - 2 
        
        print(f"Kích thước block mã hóa: {max_block_size} bytes")
        
        with open(input_file, 'rb') as f_in:
            with open(output_file, 'wb') as f_out:
                block_count = 0
                while True:
                    block = f_in.read(max_block_size)
                    if not block:
                        break
                    encrypted_block = cipher.encrypt(block)
                    
                    block_size = len(encrypted_block)
                    f_out.write(block_size.to_bytes(2, byteorder='big'))
                    f_out.write(encrypted_block)
                    
                    block_count += 1
                    if block_count % 100 == 0:
                        print(f"Đã mã hóa {block_count} blocks...")
        
        end_time = time.time()
        elapsed = end_time - start_time
        
        file_size = os.path.getsize(input_file)
        print(f"\nMã hóa thành công!")
        print(f"Số blocks: {block_count}")
        print(f"Kích thước file gốc: {file_size / (1024*1024):.2f} MB")
        print(f"Kích thước file mã hóa: {os.path.getsize(output_file) / (1024*1024):.2f} MB")
        print(f"Thời gian mã hóa: {elapsed:.4f} giây")
        
        if file_size > 0:
            print(f"Tốc độ: {(file_size / (1024*1024)) / elapsed:.2f} MB/s")
        
        return elapsed
    
    def decrypt_file(self, input_file, output_file):
        """
        Giải mã tập tin với RSA
        
        Args:
            input_file: File đầu vào (đã mã hóa)
            output_file: File đầu ra (giải mã)
        """
        if self.private_key is None:
            raise ValueError("Chưa có private key. Hãy tạo hoặc tải khóa trước.")
        
        start_time = time.time()
        cipher = PKCS1_OAEP.new(self.private_key)
        
        with open(input_file, 'rb') as f_in:
            with open(output_file, 'wb') as f_out:
                block_count = 0
                while True:
                    
                    size_bytes = f_in.read(2)
                    if not size_bytes:
                        break
                    
                    block_size = int.from_bytes(size_bytes, byteorder='big')
                    encrypted_block = f_in.read(block_size)
                    if not encrypted_block:
                        break
                    
                    decrypted_block = cipher.decrypt(encrypted_block)
                    f_out.write(decrypted_block)
                    
                    block_count += 1
                    if block_count % 100 == 0:
                        print(f"Đã giải mã {block_count} blocks...")
        
        end_time = time.time()
        elapsed = end_time - start_time
        
        file_size = os.path.getsize(output_file)
        print(f"\nGiải mã thành công!")
        print(f"Số blocks: {block_count}")
        print(f"Kích thước file: {file_size / (1024*1024):.2f} MB")
        print(f"Thời gian giải mã: {elapsed:.4f} giây")
        
        if file_size > 0:
            print(f"Tốc độ: {(file_size / (1024*1024)) / elapsed:.2f} MB/s")
        
        return elapsed


def main():
    print("=" * 60)
    print("CHƯƠNG TRÌNH MÃ HÓA/GIẢI MÃ TẬP TIN VỚI RSA")
    print("=" * 60)
    
    rsa = RSAFileEncryption()
    
    print("\nChọn chức năng:")
    print("1. Tạo cặp khóa mới")
    print("2. Mã hóa file")
    print("3. Giải mã file")
    print("4. Test hiệu năng (so sánh RSA vs DES)")
    
    choice = input("\nLựa chọn (1-4): ").strip()
    
    if choice == '1':
        # Tạo khóa
        key_size = input("Nhập kích thước khóa (mặc định 2048): ").strip()
        key_size = int(key_size) if key_size else 2048
        
        rsa.generate_keys(key_size)
        
        base_name = input("Nhập tên cơ sở cho file khóa (vd: mykey): ").strip()
        if not base_name:
            base_name = "rsa_key"
        
        rsa.save_keys(base_name)
    
    elif choice == '2':
        # Mã hóa
        pub_key_file = input("Nhập tên file public key: ").strip()
        if not os.path.exists(pub_key_file):
            print(f"File {pub_key_file} không tồn tại!")
            return
        
        rsa.load_public_key(pub_key_file)
        
        input_file = input("Nhập tên file cần mã hóa: ").strip()
        if not os.path.exists(input_file):
            print(f"File {input_file} không tồn tại!")
            return
        
        output_file = input("Nhập tên file đầu ra (mặc định: output.enc): ").strip()
        if not output_file:
            output_file = "output.enc"
        
        try:
            rsa.encrypt_file(input_file, output_file)
            print(f"\nĐã lưu file mã hóa vào: {output_file}")
        except Exception as e:
            print(f"Lỗi: {str(e)}")
            import traceback
            traceback.print_exc()
    
    elif choice == '3':
        # Giải mã
        priv_key_file = input("Nhập tên file private key: ").strip()
        if not os.path.exists(priv_key_file):
            print(f"File {priv_key_file} không tồn tại!")
            return
        
        rsa.load_private_key(priv_key_file)
        
        input_file = input("Nhập tên file cần giải mã: ").strip()
        if not os.path.exists(input_file):
            print(f"File {input_file} không tồn tại!")
            return
        
        output_file = input("Nhập tên file đầu ra (mặc định: output.dec): ").strip()
        if not output_file:
            output_file = "output.dec"
        
        try:
            rsa.decrypt_file(input_file, output_file)
            print(f"\nĐã lưu file giải mã vào: {output_file}")
        except Exception as e:
            print(f"Lỗi: {str(e)}")
            import traceback
            traceback.print_exc()
    
    elif choice == '4':
        test_comparison()
    
    else:
        print("Lựa chọn không hợp lệ!")


def test_comparison():
    """
    So sánh hiệu năng RSA vs DES với file 10MB
    """
    print("\n" + "=" * 60)
    print("SO SÁNH HIỆU NĂNG RSA vs DES")
    print("=" * 60)
    
    # Tạo file test 10MB nếu chưa có
    test_file = "test_10mb.bin"
    if not os.path.exists(test_file):
        print(f"\nTạo file test 10MB...")
        with open(test_file, 'wb') as f:
            f.write(os.urandom(10 * 1024 * 1024))
    
    # Test RSA
    print("\n--- TEST RSA ---")
    rsa = RSAFileEncryption()
    rsa.generate_keys(2048)
    
    try:
        rsa_enc_time = rsa.encrypt_file(test_file, "temp_rsa.enc")
        rsa_dec_time = rsa.decrypt_file("temp_rsa.enc", "temp_rsa.dec")
        
        os.remove("temp_rsa.enc")
        os.remove("temp_rsa.dec")
    except Exception as e:
        print(f"Lỗi RSA: {str(e)}")
        rsa_enc_time = rsa_dec_time = 0
    
    # Test DES
    print("\n--- TEST DES ---")
    try:
        from Crypto.Cipher import DES
        from Crypto.Util.Padding import pad, unpad
        
        key = b"abcdEFGH"
        
        # Mã hóa DES
        start = time.time()
        with open(test_file, 'rb') as f:
            plaintext = f.read()
        
        cipher = DES.new(key, DES.MODE_CBC)
        ciphertext = cipher.encrypt(pad(plaintext, DES.block_size))
        
        with open("temp_des.enc", 'wb') as f:
            f.write(cipher.iv)
            f.write(ciphertext)
        
        des_enc_time = time.time() - start
        print(f"DES mã hóa: {des_enc_time:.4f} giây")
        
        # Giải mã DES
        start = time.time()
        with open("temp_des.enc", 'rb') as f:
            iv = f.read(8)
            ciphertext = f.read()
        
        cipher = DES.new(key, DES.MODE_CBC, iv=iv)
        plaintext = unpad(cipher.decrypt(ciphertext), DES.block_size)
        
        with open("temp_des.dec", 'wb') as f:
            f.write(plaintext)
        
        des_dec_time = time.time() - start
        print(f"DES giải mã: {des_dec_time:.4f} giây")
        
        os.remove("temp_des.enc")
        os.remove("temp_des.dec")
    except Exception as e:
        print(f"Lỗi DES: {str(e)}")
        des_enc_time = des_dec_time = 0
    
    # So sánh
    print("\n" + "=" * 60)
    print("KẾT QUẢ SO SÁNH")
    print("=" * 60)
    print(f"\nFile test: 10MB")
    print(f"\nRSA (2048 bits):")
    print(f"  Mã hóa: {rsa_enc_time:.4f}s")
    print(f"  Giải mã: {rsa_dec_time:.4f}s")
    print(f"  Tổng: {rsa_enc_time + rsa_dec_time:.4f}s")
    
    print(f"\nDES (CBC/PKCS5):")
    print(f"  Mã hóa: {des_enc_time:.4f}s")
    print(f"  Giải mã: {des_dec_time:.4f}s")
    print(f"  Tổng: {des_enc_time + des_dec_time:.4f}s")
    
    if rsa_enc_time > 0 and des_enc_time > 0:
        print(f"\nRSA chậm hơn DES:")
        print(f"  Mã hóa: {rsa_enc_time / des_enc_time:.1f}x")
        print(f"  Giải mã: {rsa_dec_time / des_dec_time:.1f}x")


if __name__ == "__main__":
    main()

File: code/Lab05_2.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.*;
import java.security.spec.*;
import javax.crypto.*;
import java.util.Scanner;

/**
 * Lab05_2: RSA File Encryption/Decryption
 * 
 * Implements RSA encryption with:
 * - Key generation (2048-bit)
 * - Key saving/loading (DER format)
 * - File encryption/decryption with block processing
 * - Performance comparison with DES
 * 
 * Note: RSA is much slower than symmetric encryption (DES/AES).
 * For large files, consider hybrid encryption (RSA + AES).
 */
public class Lab05_2 {
    
    private KeyPair keyPair;
    private PublicKey publicKey;
    private PrivateKey privateKey;
    
    /**
     * Generates RSA key pair
     */
    public void generateKeys(int keySize) throws NoSuchAlgorithmException {
        System.out.println("Generating " + keySize + "-bit RSA key pair...");
        long startTime = System.currentTimeMillis();
        
        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
        keyGen.initialize(keySize);
        keyPair = keyGen.generateKeyPair();
        
        privateKey = keyPair.getPrivate();
        publicKey = keyPair.getPublic();
        
        long elapsed = System.currentTimeMillis() - startTime;
        System.out.println("Key generation completed in " + (elapsed / 1000.0) + " seconds");
    }
    
    /**
     * Saves keys to files in DER format
     */
    public void saveKeys(String baseFilename) throws IOException {
        // Save private key
        try (FileOutputStream fos = new FileOutputStream(baseFilename + ".key")) {
            fos.write(privateKey.getEncoded());
        }
        
        // Save public key
        try (FileOutputStream fos = new FileOutputStream(baseFilename + ".pub")) {
            fos.write(publicKey.getEncoded());
        }
        
        System.out.println("Keys saved:");
        System.out.println("  Private key: " + baseFilename + ".key");
        System.out.println("  Public key: " + baseFilename + ".pub");
    }
    
    /**
     * Loads public key from file
     */
    public void loadPublicKey(String filename) throws Exception {
        byte[] keyBytes = Files.readAllBytes(Paths.get(filename));
        X509EncodedKeySpec spec = new X509EncodedKeySpec(keyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        publicKey = keyFactory.generatePublic(spec);
        System.out.println("Loaded public key from " + filename);
    }
    
    /**
     * Loads private key from file
     */
    public void loadPrivateKey(String filename) throws Exception {
        byte[] keyBytes = Files.readAllBytes(Paths.get(filename));
        PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        privateKey = keyFactory.generatePrivate(spec);
        System.out.println("Loaded private key from " + filename);
    }
    
    /**
     * Encrypts a file using RSA with block processing
     * 
     * RSA can only encrypt data smaller than key size, so we process in blocks.
     * Each encrypted block is prefixed with its size (2 bytes).
     */
    public double encryptFile(String inputFile, String outputFile) throws Exception {
        if (publicKey == null) {
            throw new IllegalStateException("No public key loaded. Generate or load a key first.");
        }
        
        long startTime = System.nanoTime();
        
        Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
        cipher.init(Cipher.ENCRYPT_MODE, publicKey);
        
        // Calculate max block size for RSA encryption
        int keySizeBytes = ((java.security.interfaces.RSAPublicKey) publicKey).getModulus().bitLength() / 8;
        int maxBlockSize = keySizeBytes - 11;  // PKCS1Padding overhead is 11 bytes
        
        System.out.println("Max block size for encryption: " + maxBlockSize + " bytes");
        
        try (FileInputStream fis = new FileInputStream(inputFile);
             FileOutputStream fos = new FileOutputStream(outputFile)) {
            
            byte[] buffer = new byte[maxBlockSize];
            int blockCount = 0;
            int bytesRead;
            
            while ((bytesRead = fis.read(buffer)) != -1) {
                // Encrypt block
                byte[] inputBlock = new byte[bytesRead];
                System.arraycopy(buffer, 0, inputBlock, 0, bytesRead);
                byte[] encryptedBlock = cipher.doFinal(inputBlock);
                
                // Write block size (2 bytes) then encrypted block
                int blockSize = encryptedBlock.length;
                fos.write((blockSize >> 8) & 0xFF);
                fos.write(blockSize & 0xFF);
                fos.write(encryptedBlock);
                
                blockCount++;
                if (blockCount % 100 == 0) {
                    System.out.println("Encrypted " + blockCount + " blocks...");
                }
            }
            
            long endTime = System.nanoTime();
            double elapsed = (endTime - startTime) / 1_000_000_000.0;
            
            // Print statistics
            File inFile = new File(inputFile);
            File outFile = new File(outputFile);
            double inSizeMB = inFile.length() / (1024.0 * 1024.0);
            double outSizeMB = outFile.length() / (1024.0 * 1024.0);
            
            System.out.println("\nEncryption successful!");
            System.out.println("Blocks processed: " + blockCount);
            System.out.println(String.format("Original file size: %.2f MB", inSizeMB));
            System.out.println(String.format("Encrypted file size: %.2f MB", outSizeMB));
            System.out.println(String.format("Encryption time: %.4f seconds", elapsed));
            if (elapsed > 0) {
                System.out.println(String.format("Speed: %.2f MB/s", inSizeMB / elapsed));
            }
            
            return elapsed;
        }
    }
    
    /**
     * Decrypts a file using RSA with block processing
     */
    public double decryptFile(String inputFile, String outputFile) throws Exception {
        if (privateKey == null) {
            throw new IllegalStateException("No private key loaded. Generate or load a key first.");
        }
        
        long startTime = System.nanoTime();
        
        Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
        cipher.init(Cipher.DECRYPT_MODE, privateKey);
        
        try (FileInputStream fis = new FileInputStream(inputFile);
             FileOutputStream fos = new FileOutputStream(outputFile)) {
            
            int blockCount = 0;
            
            while (true) {
                // Read block size (2 bytes)
                int byte1 = fis.read();
                int byte2 = fis.read();
                
                if (byte1 == -1 || byte2 == -1) {
                    break;  // End of file
                }
                
                int blockSize = ((byte1 & 0xFF) << 8) | (byte2 & 0xFF);
                
                // Read encrypted block
                byte[] encryptedBlock = new byte[blockSize];
                int bytesRead = fis.read(encryptedBlock);
                
                if (bytesRead != blockSize) {
                    throw new IOException("Unexpected end of file");
                }
                
                // Decrypt block
                byte[] decryptedBlock = cipher.doFinal(encryptedBlock);
                fos.write(decryptedBlock);
                
                blockCount++;
                if (blockCount % 100 == 0) {
                    System.out.println("Decrypted " + blockCount + " blocks...");
                }
            }
            
            long endTime = System.nanoTime();
            double elapsed = (endTime - startTime) / 1_000_000_000.0;
            
            // Print statistics
            File outFile = new File(outputFile);
            double outSizeMB = outFile.length() / (1024.0 * 1024.0);
            
            System.out.println("\nDecryption successful!");
            System.out.println("Blocks processed: " + blockCount);
            System.out.println(String.format("Decrypted file size: %.2f MB", outSizeMB));
            System.out.println(String.format("Decryption time: %.4f seconds", elapsed));
            if (elapsed > 0) {
                System.out.println(String.format("Speed: %.2f MB/s", outSizeMB / elapsed));
            }
            
            return elapsed;
        }
    }
    
    /**
     * Interactive main menu
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Lab05_2 rsa = new Lab05_2();
        
        System.out.println("============================================================");
        System.out.println("RSA FILE ENCRYPTION/DECRYPTION PROGRAM");
        System.out.println("============================================================");
        
        try {
            System.out.println("\nChoose function:");
            System.out.println("1. Generate new key pair");
            System.out.println("2. Encrypt file");
            System.out.println("3. Decrypt file");
            System.out.println("4. Performance test (compare RSA vs DES)");
            
            System.out.print("\nChoice (1-4): ");
            String choice = scanner.nextLine().trim();
            
            switch (choice) {
                case "1":
                    // Generate keys
                    System.out.print("Enter key size (default 2048): ");
                    String sizeStr = scanner.nextLine().trim();
                    int keySize = sizeStr.isEmpty() ? 2048 : Integer.parseInt(sizeStr);
                    
                    rsa.generateKeys(keySize);
                    
                    System.out.print("Enter base name for key files (e.g., mykey): ");
                    String baseName = scanner.nextLine().trim();
                    if (baseName.isEmpty()) {
                        baseName = "rsa_key";
                    }
                    
                    rsa.saveKeys(baseName);
                    break;
                    
                case "2":
                    // Encrypt
                    System.out.print("Enter public key file name: ");
                    String pubKeyFile = scanner.nextLine().trim();
                    
                    if (!new File(pubKeyFile).exists()) {
                        System.out.println("File " + pubKeyFile + " does not exist!");
                        return;
                    }
                    
                    rsa.loadPublicKey(pubKeyFile);
                    
                    System.out.print("Enter file to encrypt: ");
                    String inputFile = scanner.nextLine().trim();
                    
                    if (!new File(inputFile).exists()) {
                        System.out.println("File " + inputFile + " does not exist!");
                        return;
                    }
                    
                    System.out.print("Enter output file name (default: output.enc): ");
                    String outputFile = scanner.nextLine().trim();
                    if (outputFile.isEmpty()) {
                        outputFile = "output.enc";
                    }
                    
                    rsa.encryptFile(inputFile, outputFile);
                    System.out.println("\nFile saved to: " + outputFile);
                    break;
                    
                case "3":
                    // Decrypt
                    System.out.print("Enter private key file name: ");
                    String privKeyFile = scanner.nextLine().trim();
                    
                    if (!new File(privKeyFile).exists()) {
                        System.out.println("File " + privKeyFile + " does not exist!");
                        return;
                    }
                    
                    rsa.loadPrivateKey(privKeyFile);
                    
                    System.out.print("Enter file to decrypt: ");
                    String encFile = scanner.nextLine().trim();
                    
                    if (!new File(encFile).exists()) {
                        System.out.println("File " + encFile + " does not exist!");
                        return;
                    }
                    
                    System.out.print("Enter output file name (default: output.dec): ");
                    String decFile = scanner.nextLine().trim();
                    if (decFile.isEmpty()) {
                        decFile = "output.dec";
                    }
                    
                    rsa.decryptFile(encFile, decFile);
                    System.out.println("\nFile saved to: " + decFile);
                    break;
                    
                case "4":
                    // Performance test
                    performanceTest();
                    break;
                    
                default:
                    System.out.println("Invalid choice!");
            }
            
        } catch (Exception e) {
            System.out.println("\nError: " + e.getMessage());
            e.printStackTrace();
        } finally {
            scanner.close();
        }
    }
    
    /**
     * Performance comparison between RSA and DES
     */
    private static void performanceTest() throws Exception {
        System.out.println("\n============================================================");
        System.out.println("PERFORMANCE COMPARISON: RSA vs DES");
        System.out.println("============================================================");
        
        // Create test file if needed
        String testFile = "test_10mb.bin";
        if (!new File(testFile).exists()) {
            System.out.println("\nCreating 10MB test file...");
            try (FileOutputStream fos = new FileOutputStream(testFile)) {
                byte[] buffer = new byte[1024 * 1024];  // 1MB buffer
                SecureRandom random = new SecureRandom();
                for (int i = 0; i < 10; i++) {
                    random.nextBytes(buffer);
                    fos.write(buffer);
                }
            }
            System.out.println("Test file created: " + testFile);
        }
        
        // Test RSA
        System.out.println("\n--- Testing RSA ---");
        Lab05_2 rsa = new Lab05_2();
        rsa.generateKeys(2048);
        
        double rsaEncTime = rsa.encryptFile(testFile, "temp_rsa.enc");
        double rsaDecTime = rsa.decryptFile("temp_rsa.enc", "temp_rsa.dec");
        
        new File("temp_rsa.enc").delete();
        new File("temp_rsa.dec").delete();
        
        // Test DES
        System.out.println("\n--- Testing DES ---");
        byte[] desKey = "abcdEFGH".getBytes(StandardCharsets.UTF_8);
        
        double desEncTime = Lab05_1.encryptFile(testFile, "temp_des.enc", desKey, "CBC", "PKCS5Padding");
        double desDecTime = Lab05_1.decryptFile("temp_des.enc", "temp_des.dec", desKey, "CBC", "PKCS5Padding");
        
        new File("temp_des.enc").delete();
        new File("temp_des.dec").delete();
        
        // Print comparison
        System.out.println("\n============================================================");
        System.out.println("COMPARISON RESULTS");
        System.out.println("============================================================");
        System.out.println("\nTest file: 10MB");
        System.out.println("\nRSA (2048-bit):");
        System.out.println(String.format("  Encryption: %.4f seconds", rsaEncTime));
        System.out.println(String.format("  Decryption: %.4f seconds", rsaDecTime));
        System.out.println(String.format("  Total: %.4f seconds", rsaEncTime + rsaDecTime));
        
        System.out.println("\nDES (CBC/PKCS5Padding):");
        System.out.println(String.format("  Encryption: %.4f seconds", desEncTime));
        System.out.println(String.format("  Decryption: %.4f seconds", desDecTime));
        System.out.println(String.format("  Total: %.4f seconds", desEncTime + desDecTime));
        
        if (desEncTime > 0 && desDecTime > 0) {
            System.out.println("\nRSA is slower than DES:");
            System.out.println(String.format("  Encryption: %.1fx", rsaEncTime / desEncTime));
            System.out.println(String.format("  Decryption: %.1fx", rsaDecTime / desDecTime));
        }
    }
}
